src/zcash/JoinSplit.cpp

```diff
@@ -0,0 +1,344 @@
+        for (size_t i = 0; i < NumOutputs; i++) {
+            commitments[i] = output_notes[i].cm();
+        }
+
+        // Encrypt the ciphertexts containing the note
+        // plaintexts to the recipients of the value.
+        {
+            ZCNoteEncryption encryptor(h_sig);
+
+            for (size_t i = 0; i < NumOutputs; i++) {
+                // TODO: expose memo in the public interface
+                boost::array<unsigned char, ZC_MEMO_SIZE> memo;
+                memo[0] = 0xF6; // invalid UTF8 as per spec
+
+                NotePlaintext pt(output_notes[i], memo);
+
+                ciphertexts[i] = pt.encrypt(encryptor, outputs[i].addr.pk_enc);
+            }
+
+            ephemeralKey = encryptor.get_epk();
+        }
+
+        // Authenticate h_sig with each of the input
+        // spending keys, producing macs which protect
+        // against malleability.
+        for (size_t i = 0; i < NumInputs; i++) {
+            macs[i] = PRF_pk(inputs[i].key, i, h_sig);
+        }
+
+        std::vector<FieldT> primary_input;
+        std::vector<FieldT> aux_input;
+
+        {
+            protoboard<FieldT> pb;
+            {
+                joinsplit_gadget<FieldT, NumInputs, NumOutputs> g(pb);
+                g.generate_r1cs_constraints();
+                g.generate_r1cs_witness(
+                    phi,
+                    rt,
+                    h_sig,
+                    inputs,
+                    output_notes,
+                    vpub_old,
+                    vpub_new
+                );
+            }
+
+            if (!pb.is_satisfied()) {
+                throw std::invalid_argument("Constraint system not satisfied by inputs");
+            }
+
+            primary_input = pb.primary_input();
+            aux_input = pb.auxiliary_input();
+        }
+
+        auto proof = r1cs_ppzksnark_prover<ppzksnark_ppT>(
+            *pk,
+            primary_input,
+            aux_input
+        );
+
+        std::stringstream ss;
+        ss << proof;
+
+        return ss.str();
+    }
+};
+
+template<size_t NumInputs, size_t NumOutputs>
+JoinSplit<NumInputs, NumOutputs>* JoinSplit<NumInputs, NumOutputs>::Generate()
+{
+    JoinSplitCircuit<NumInputs, NumOutputs>::initialize();
+    auto js = new JoinSplitCircuit<NumInputs, NumOutputs>();
+    js->generate();
+
+    return js;
+}
+
+template<size_t NumInputs, size_t NumOutputs>
+JoinSplit<NumInputs, NumOutputs>* JoinSplit<NumInputs, NumOutputs>::Unopened()
+{
+    JoinSplitCircuit<NumInputs, NumOutputs>::initialize();
+    return new JoinSplitCircuit<NumInputs, NumOutputs>();
+}
+
+template<size_t NumInputs, size_t NumOutputs>
+uint256 JoinSplit<NumInputs, NumOutputs>::h_sig(
+    const uint256& randomSeed,
+    const boost::array<uint256, NumInputs>& nullifiers,
+    const uint256& pubKeyHash
+) {
+    unsigned char personalization[crypto_generichash_blake2b_PERSONALBYTES]
+        = {'Z','c','a','s','h','C','o','m','p','u','t','e','h','S','i','g'};
```

Comment:
is strlen("ZcashComputehSig") == crypto_generichash_blake2b_PERSONALBYTES? Yes it is.